src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicTlsSecret.cs

```diff
@@ -19,88 +20,83 @@ internal sealed class MsQuicTlsSecret : IDisposable
-                }
-
-                if (_tlsSecrets->IsSet.ServerTrafficSecret0 != 0)
-                {
-                    stream.Write(Encoding.ASCII.GetBytes($"SERVER_TRAFFIC_SECRET_0 {clientRandom} {HexConverter.ToString(new ReadOnlySpan<byte>(_tlsSecrets->ServerTrafficSecret0, _tlsSecrets->SecretLength))}\n"));
-                }
-
-                if (_tlsSecrets->IsSet.ClientEarlyTrafficSecret != 0)
-                {
-                    stream.Write(Encoding.ASCII.GetBytes($"CLIENT_EARLY_TRAFFIC_SECRET {clientRandom} {HexConverter.ToString(new ReadOnlySpan<byte>(_tlsSecrets->ClientEarlyTrafficSecret, _tlsSecrets->SecretLength))}\n"));
-                }
-
-                stream.Flush();
+                clientRandom = HexConverter.ToString(new ReadOnlySpan<byte>(_tlsSecrets->ClientRandom, 32));
+            }
+            if (_tlsSecrets->IsSet.ClientHandshakeTrafficSecret != 0)
+            {
+                s_fileStream.Write(Encoding.ASCII.GetBytes($"CLIENT_HANDSHAKE_TRAFFIC_SECRET {clientRandom} {HexConverter.ToString(new ReadOnlySpan<byte>(_tlsSecrets->ClientHandshakeTrafficSecret, _tlsSecrets->SecretLength))}\n"));
+            }
+            if (_tlsSecrets->IsSet.ServerHandshakeTrafficSecret != 0)
+            {
+                s_fileStream.Write(Encoding.ASCII.GetBytes($"SERVER_HANDSHAKE_TRAFFIC_SECRET {clientRandom} {HexConverter.ToString(new ReadOnlySpan<byte>(_tlsSecrets->ServerHandshakeTrafficSecret, _tlsSecrets->SecretLength))}\n"));
             }
+            if (_tlsSecrets->IsSet.ClientTrafficSecret0 != 0)
+            {
+                s_fileStream.Write(Encoding.ASCII.GetBytes($"CLIENT_TRAFFIC_SECRET_0 {clientRandom} {HexConverter.ToString(new ReadOnlySpan<byte>(_tlsSecrets->ClientTrafficSecret0, _tlsSecrets->SecretLength))}\n"));
+            }
+            if (_tlsSecrets->IsSet.ServerTrafficSecret0 != 0)
+            {
+                s_fileStream.Write(Encoding.ASCII.GetBytes($"SERVER_TRAFFIC_SECRET_0 {clientRandom} {HexConverter.ToString(new ReadOnlySpan<byte>(_tlsSecrets->ServerTrafficSecret0, _tlsSecrets->SecretLength))}\n"));
+            }
+            if (_tlsSecrets->IsSet.ClientEarlyTrafficSecret != 0)
+            {
+                s_fileStream.Write(Encoding.ASCII.GetBytes($"CLIENT_EARLY_TRAFFIC_SECRET {clientRandom} {HexConverter.ToString(new ReadOnlySpan<byte>(_tlsSecrets->ClientEarlyTrafficSecret, _tlsSecrets->SecretLength))}\n"));
+            }
+            s_fileStream.Flush();
         }
+
+        NativeMemory.Clear(_tlsSecrets, (nuint)sizeof(QUIC_TLS_SECRETS));
     }
 
     public unsafe void Dispose()
     {
-        if (_tlsSecrets != null)
+        if (_tlsSecrets is not null)
         {
-            NativeMemory.Clear(_tlsSecrets, (nuint)sizeof(QUIC_TLS_SECRETS));
+            NativeMemory.Free(_tlsSecrets);
```

Comment:
I'm not sure this is safe. It feels like we can free the memory while MsQuic still may be trying to write to that. That was one reason why originally the native memory was tight to the MsQuic handle lifecycle.

Reply:
The dispose is called when the connection handle gets released, not from `QuicConnection.DisposeAsync`.